### PR TITLE
Change interactive window behavior when typing inside readonly area

### PR DESCRIPTION
**Pull request by genlu**
Fix issue https://github.com/dotnet/roslyn/issues/5130

Here's the new behavior for all deletion without selected text related operations:
- if caret is in editable area, no change in behavior.
- if caret is in active prompt area, move caret to the closest location first.
- if caret is in other readonly area, do nothing.

Here's the new behavior for all deletion with selected text related operations:
- if the selection overlaps with editable area, caret will be moved to a position in
  editable area that is closest to it's original position.
- Otherwise, do nothing (and preserve the selection).

Deletion related operations including:
- `Delete` (and TypeChar)
- `Backspace`
- `Return` and `BreakLine`
- `Paste`
- `Cut`

Question:
`InsertCode` method also depends on `CutOrDeleteSelection` method, but I'm not sure if this change will affect it, can I assume the caret will always in editable area when `InsertCode` is called?

### CODE REVIEW

**Comment by genlu** [comment]
@amcasey @tmat @cston @KevinH-MS @ManishJayaswal Please review.

This is not completed, I will add tests in later commits. Just wanna send out a PR so we can discuss if this new behavior makes sense.

**Comment by genlu** [comment]
@tmat Have you reviewed the lastest commit as well? I have changed the behavior of `CutOrDeleteSelection` method,please let me know if it might break anything.

**Comment by tmat** [comment]
Looks reasonable. Would need to play with it to see if everything is actually good.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 1]
```diff
@@ -577,7 +577,23 @@ private void RestoreUncommittedInput()
             /// </summary>
             public bool Paste()
             {
-                MoveCaretToClosestEditableBuffer();
+                if (!TextView.Selection.IsEmpty)
+                {
+                    if (CutOrDeleteSelection(isCut: false))
+                    {
+                        MoveCaretToClosestEditableBuffer();
+                    }
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Else return false?

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 2]
```diff
@@ -577,7 +577,23 @@ private void RestoreUncommittedInput()
...
+                }
+                else if (IsInActivePrompt(TextView.Caret.Position.BufferPosition))
+                {
+                    MoveCaretToClosestEditableBuffer();
+                }
+
+                // selection isn't empty means none of the selected area is editable
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Unnecessary after above else.

Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Personally, I would expect this method to return false if the caret is in a readonly region, but no one actually consumes the return value so it's not urgent.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 3]
```diff
@@ -1148,6 +1164,32 @@ private int GetPromptIndexForPoint(ReadOnlyCollection<SnapshotSpan> sourceSpans,
                 }
                 return index;
             }
+            
+            private bool IsInActivePrompt(SnapshotPoint point)
+            {
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: I think you could use ```TextView.BufferGraph.MapDownToFirst``` to find the subject buffer containing the point and then just check if it's ```editableBuffer``` (which I would call ```activeBuffer``` for consistency with the method name.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 4]
```diff
@@ -2103,17 +2145,27 @@ public bool Delete()
...
                 if (!TextView.Selection.IsEmpty)
                 {
-                    if (TextView.Selection.Mode == TextSelectionMode.Stream || ReduceBoxSelectionToEditableBox())
+                    if (CutOrDeleteSelection(isCut: false))
                     {
-                        CutOrDeleteSelection(isCut: false);
                         MoveCaretToClosestEditableBuffer();
-                        handled = true;
                     }
+                    handled = true;
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Why return true when ```CutOrDeleteSelection``` fails?

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 5]
```diff
@@ -2103,17 +2145,27 @@ public bool Delete()
...
+                }
+                else if (IsInActivePrompt(TextView.Caret.Position.BufferPosition))
+                {
+                    MoveCaretToClosestEditableBuffer();
                 }
-
                 return handled;
             }
 
+            private bool IsSteamSelectionInEditableBuffer(ITextSelection selection)
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Typo "Steam".

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 6]
```diff
@@ -2186,8 +2238,7 @@ private bool ReduceBoxSelectionToEditableBox(ref VirtualSnapshotPoint selectionT
                     if (selectionLeftColumn > maxPromptLength || maxPromptLength == minPromptLength)
                     {
                         selectionTopLine = editableLine;
-                        selectionLeftColumn = Math.Max(selectionLeftColumn, maxPromptLength);
-                        result = false;
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: I can't tell what this changes.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 3 above)
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Even better, see ```GetPositionInLanguageBuffer``` and ```GetPositionInStandardInputBuffer```.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 7]
```diff
@@ -2420,13 +2489,19 @@ public bool Backspace()
                 {
                     if (TextView.Selection.Mode == TextSelectionMode.Stream || ReduceBoxSelectionToEditableBox())
                     {
-                        CutOrDeleteSelection(isCut: false);
-                        MoveCaretToClosestEditableBuffer();
+                        if (CutOrDeleteSelection(isCut: false))
+                        {
+                            MoveCaretToClosestEditableBuffer();
+                        }
                         handled = true;
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Only handled if delete succeeds?

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 8]
```diff
@@ -2580,8 +2655,31 @@ private bool HandlePostServicesReturn(bool trySubmit)
                     return false;
                 }
 
+                if (!TextView.Selection.IsEmpty)
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Same comments as in ```Paste```.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 6 above)
Comment (genlu) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: My understanding is `ReduceBoxSelectionToEditableBox` returns `false` if the selection can not be reduced to editable box, and it is not true in this case, therefore `result = false;` is removed.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 1 above)
Comment (genlu) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Fixed.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 2 above)
Comment (genlu) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Fixed.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 5 above)
Comment (genlu) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Fixed.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 4 above)
Comment (genlu) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Fixed.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 7 above)
Comment (genlu) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Fixed.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 4 above)
Comment (genlu) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: @amcasey Not sure I understand your comment here. This method tries to decide if the given point lies with in the span of an active prompt, not a editable buffer (my understanding is all the prompts are in a single readonly buffer, which span over multiple spans).

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 8 above)
Comment (genlu) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Fixed.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 3 above)
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: @genlu Yes, I appear to have misunderstood what the method was trying to do.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 9]
```diff
@@ -1148,6 +1164,32 @@ private int GetPromptIndexForPoint(ReadOnlyCollection<SnapshotSpan> sourceSpans,
...
+                if (editableBuffer == null)
+                {
+                    return false;
+                }
+                                                                           
+                var sourceSpans = GetSourceSpans(point.Snapshot);
+                var index = GetSourceSpanIndex(sourceSpans, point);
+                if (index == sourceSpans.Count)
+                {
+                    index--;
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: ```return false;```?

Comment (genlu) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: @amcasey Well, `GetSourceSpanIndex` returns the length of the collection if the point is at the end of the last span. Are you suggesting that the prompt will never be in the last span?

Comment (amcasey) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: That's true in practice, but we probably don't want to depend on it.  I'm very surprised to learn that ``` GetSourceSpanIndex``` behaves that way, but it looks like all callers handle it.

**Comment by amcasey** [comment]
:+1:

**Comment by tannergooding** [comment]
test full please

**Comment by genlu** [comment]
FYI @ManishJayaswal @kuhlenh @tmat @amcasey @cston @KevinH-MS 
If you found any of the new behaviors doesn't make sense, please let me know.

**Pull request merged**